core/rawdb/accessors_chain.go

```diff
@@ -269,6 +269,49 @@ func WriteFastTxLookupLimit(db ethdb.KeyValueWriter, number uint64) {
 	}
 }
 
+// ReadHeadersRLP returns the rlp-encoded headers, starting at 'number', and going
+// backwards towards genesis. This method assumes that the caller already has
+// placed a cap on count, to prevent DoS issues.
+func ReadHeadersRLP(db ethdb.Reader, number uint64, count uint64) []rlp.RawValue {
+	var rlpHeaders []rlp.RawValue
+	if count == 0 {
+		return rlpHeaders
+	}
+	i := number
+	if count-1 > number {
+		// It's ok to request block 0, 1 item
+		count = number + 1
+	}
+	limit, _ := db.Ancients()
+	// First read live blocks
+	if i >= limit {
+		// If we need to read live blocks, we need to figure out the hash first
+		hash := ReadCanonicalHash(db, number)
+		var hdr types.Header
+		for ; i >= limit && count > 0; i-- {
+			if data, _ := db.Get(headerKey(i, hash)); len(data) > 0 {
+				rlpHeaders = append(rlpHeaders, data)
+				// Get the parent hash for next query
+				rlp.DecodeBytes(data, &hdr)
```

Comment:
So there is a hack to extract the parent hash directly from rlp-encoded form:

- It seems `parentHash` is the first field
- So we only need to know how big is the rlp-prefix. I'm guessing that should be 3 bytes. And even though header size is not fixed I don't think the difference is so big as to need an extra length byte.
- If I'm not wrong then we can get it by `data[3:37]`

edit: the hash itself is also encoded and has a prefix (1-2 bytes?), so there's that too

Reply:
Awesome, thanks, fixed!